### PR TITLE
Add API for License templates

### PR DESCRIPTION
**Pull request by JakobDev**
This adds a API for getting License templates.  This tries to be as close to the [GitHub API](https://docs.github.com/en/rest/licenses?apiVersion=2022-11-28) as possible, but Gitea does not support all features that GitHub has. I think they should been added, but this out f the scope of this PR. You should merge #23006 before this PR for security reasons.

### CODE REVIEW

File: routers/api/v1/misc/licenses.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	//   type: string
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/LicenseTemplateInfo"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+	name := ctx.Params("name")
+
+	text, err := options.License(name)
```
Comment (wxiaoguang) [inline routers/api/v1/misc/licenses.go:60]: The same as https://github.com/go-gitea/gitea/pull/22783/files#r1175127251

File: routers/api/v1/misc/licenses.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	if err != nil {
+		ctx.NotFound()
+		return
+	}
+
+	response := api.LicenseTemplateInfo{
+		Key:  name,
+		Name: name,
+		URL:  fmt.Sprintf("%sapi/v1/licenses/%s", setting.AppURL, name),
```
Comment (wxiaoguang) [inline routers/api/v1/misc/licenses.go]: To use a variable in URL, it needs to be escaped.

File: routers/api/v1/misc/licenses.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	// - application/json
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/LicenseTemplateList"
+	response := make([]api.LicensesTemplateListEntry, len(repo_module.Licenses))
+	for i, license := range repo_module.Licenses {
+		response[i] = api.LicensesTemplateListEntry{
+			Key:  util.PathJoinRelX(license),
+			Name: util.PathJoinRelX(license),
+			URL:  fmt.Sprintf("%sapi/v1/licenses/%s", setting.AppURL, html.EscapeString(license)),
```
Comment (wxiaoguang) [inline routers/api/v1/misc/licenses.go:34]: Oh sorry, not `html.EscapeString`, but `url.QueryEscape()`

Comment (wxiaoguang) [reply routers/api/v1/misc/licenses.go:34]: Oh sorry, not `html.EscapeString`, but `url.PathEscape()`

Comment (wxiaoguang) [inline routers/api/v1/misc/licenses.go]: Oh sorry, not `html.EscapeString`, but `url.PathEscape()`

(sorry for bothering, GitHub web UI got some mess with updating comment ....)

File: routers/api/v1/misc/licenses.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	// produces:
+	// - application/json
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/LicenseTemplateList"
+	response := make([]api.LicensesTemplateListEntry, len(repo_module.Licenses))
+	for i, license := range repo_module.Licenses {
+		response[i] = api.LicensesTemplateListEntry{
+			Key:  util.PathJoinRelX(license),
+			Name: util.PathJoinRelX(license),
```
Comment (wxiaoguang) [inline routers/api/v1/misc/licenses.go]:
```suggestion
			Key:  license,
			Name: license,
```

**Review by wxiaoguang — APPROVED** [review]
`util.PathJoinRelX` means "clean the path to a relative path", it's only needed for uncontrolable user input. It has no affect on Gitea internal data like `Licenses`.

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
> `util.PathJoinRelX` means "clean the path to a relative path", it's only needed for uncontrolable user input. It has no affect on Gitea internal data like `Licenses`.

Is that a open request?

**Comment by silverwind** [comment]
Fixed conflict and squashed.

**Comment by wxiaoguang** [comment]
> > `util.PathJoinRelX` means "clean the path to a relative path", it's only needed for uncontrolable user input. It has no affect on Gitea internal data like `Licenses`.
> 
> Is that a open request?

Nope, already resolved.

**Pull request merged**